### DISCUSSION TITLE
Use services in some jobs of a matrix

### DISCUSSION THREAD

**Discussion by datapythonista** [community/community#25578] [Actions]
<p>Hi!</p>
<p>I’ve got a complex job, consisting of many steps that I want to run in Linux and Windows. The only difference, besides the OS, is that in Linux there is a step using PostgreSQL, and in Windows this is skipped.</p>
<p>I’m able to have the next configuration yaml, that is exactly what I want, except that the postgres service should not start for Windows, and it does. Not sure if postgres is able to start on windows, but in practice I need more services (MySQL…) and some of them surely fail to start in Windows.</p>
<pre><code>jobs:
  complex_job:
    name: Complex job
    runs-on: ${{ matrix.os }}
    services:
      postgres:
        image: postgres:latest
    strategy:
      matrix:
        os: [ubuntu-latest, windows-latest]
    steps:
      # Lots of steps that I don't want to duplicate
</code></pre>
<p>I can make this work if I create two jobs, one for Windows without the services, and one for Linux with the services. But I need to copy the long list of steps, and make all changes twice in the future.</p>
<p>Is there a way to do what I want in a reasonable way?</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi,</p>
<p>Thanks for your feedback!</p>
<p>It’s a limitation to use services in some jobs of matrix. ‘<a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idservices" rel="nofollow noopener">jobs.&lt;job_id&gt;.services</a>’ has the same level with <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idstrategy" rel="nofollow noopener">jobs.&lt;job_id&gt;.strategy</a>, and it will initialize the container before the steps execution. Please create individul jobs for different runner os.</p>
